### ISSUE TITLE
Managing b2World state for rollback

### ISSUE THREAD

**Issue by briansemrau** [briansemrau/godot_box2d#66]
I'm looking into server-side rollback and resimulation in a project.

To set the b2World to a prior state, I am under the impression that it isn't a great idea to manually set the transforms, velocities, etc. to the stored prior state. I'm not certain how important the b2 internal states are. Would it create brief non-physical behavior? (It certainly would be slow, teleporting everything around and rebuilding the internal state.) 

If my understanding is correct, then it would be valuable to be able to essentially memcpy the b2World. Maybe you can already deep-copy parts of the scene tree, I don't know.

---

Another thing that comes to mind is optimization. It would make sense to only rollback the relevant islands, not the whole world*. Depending on how frequent rollback/resim occurs, this could create a significant performance improvement if possible. The problem is, I don't know how feasible this is to implement, especially in the context of this module.

\* on second thought, this is a bad idea. Islands that were previously irrelevant may become relevant. There are ways to handle this, but it's complicated

**Comment by jordo** [comment]
this is going to be VERY context dependent.  as there are a lot of different ways to do reset/rollback & re-sim, all which have trade-offs in a box2d context.

Which requirements a particular game has (contact persistence, how much determinism, etc) is going to be a big factor so I don't think there is a silver bullet that be built to handle a general case very well.   I'm really unsure about supporting anything related to this in the module itself.

We manually set transforms, velocities, etc, but a good idea in that case is to remove bodies from world temporarily, and re-insert them in the same order after positions & velocities have been updated.  (Resetting to a known state, so no temporary overlaps during a reset).    Same-order insertion is important as fixtures and bodies in box2d are a list, and same order means the solver will act the same way post reset.  

The performance hit to set transfom in a reset is FindNewContacts, but there isn't much way around that if you do things this way.  The nice thing, is if re-simulating say 10 ticks, FindNewContacts is still only run once on the first set_Transform().

**Comment by briansemrau** [comment]
I appreciate your input on this. It's hard to find information on this subject online.

Just because it's fun to look at, I want to share something I found that looks like a general solution if you work with it correctly but it's very hacky:
https://github.com/ixchow/Checkpointable-Box2D
It appears to work by hijacking Box2D memory management and storing/resetting the memory blocks. Looks ripe for opportunity for memory corruption :D

I've also realized some of the challenges of integrating this with Godot itself, such as handling Box2D nodes that have been removed since the last stored state. I don't think there's a seamless way to make this work without exposing application developers to many strange bugs.

With all of this considered, I'll agree with you that this would be hard to support module-side.
